overlord/devicestate/devicestate_remodel_test.go

```diff
@@ -1623,3 +1648,170 @@ func (s *deviceMgrSuite) TestRemodelSwitchBase(c *C) {
+				"presence": "required",
+			},
+			map[string]interface{}{
+				"name":     "new-required-snap-2",
+				"id":       snaptest.AssertedSnapID("new-required-snap-2"),
+				"presence": "required",
+			},
+		},
+	})
+	chg, err := devicestate.Remodel(s.state, new)
+	c.Assert(err, IsNil)
+	c.Assert(chg.Summary(), Equals, "Refresh model assertion from revision 0 to 1")
+
+	tl := chg.Tasks()
+	// 2 snaps (3 tasks for each) + recovery system (2 tasks) + set-model
+	c.Assert(tl, HasLen, 2*3+2+1)
+
+	deviceCtx, err := devicestate.DeviceCtx(s.state, tl[0], nil)
+	c.Assert(err, IsNil)
+	// deviceCtx is actually a remodelContext here
+	remodCtx, ok := deviceCtx.(devicestate.RemodelContext)
+	c.Assert(ok, Equals, true)
+	c.Check(remodCtx.ForRemodeling(), Equals, true)
+	c.Check(remodCtx.Kind(), Equals, devicestate.UpdateRemodel)
+	c.Check(remodCtx.Model(), DeepEquals, new)
+	c.Check(remodCtx.Store(), IsNil)
+
+	// check the tasks
+	tDownloadSnap1 := tl[0]
+	tValidateSnap1 := tl[1]
+	tInstallSnap1 := tl[2]
+	tDownloadSnap2 := tl[3]
+	tValidateSnap2 := tl[4]
+	tInstallSnap2 := tl[5]
+	tCreateRecovery := tl[6]
+	tFinalizeRecovery := tl[7]
+	tSetModel := tl[8]
+
+	// check the tasks
+	c.Assert(tDownloadSnap1.Kind(), Equals, "fake-download")
+	c.Assert(tDownloadSnap1.Summary(), Equals, "Download new-required-snap-1")
+	c.Assert(tDownloadSnap1.WaitTasks(), HasLen, 0)
+	c.Assert(tValidateSnap1.Kind(), Equals, "validate-snap")
+	c.Assert(tValidateSnap1.Summary(), Equals, "Validate new-required-snap-1")
+	expectedLabel := now.Format("20060102")
+	c.Assert(tCreateRecovery.Kind(), Equals, "create-recovery-system")
+	c.Assert(tCreateRecovery.Summary(), Equals, fmt.Sprintf("Create recovery system with label %q", expectedLabel))
+	c.Assert(tFinalizeRecovery.Kind(), Equals, "finalize-recovery-system")
+	c.Assert(tFinalizeRecovery.Summary(), Equals, fmt.Sprintf("Finalize recovery system with label %q", expectedLabel))
+	// check the ordering, download/validate everything first, then install
+
+	// snap2 downloads wait for the downloads of snap1
+	c.Assert(tDownloadSnap1.WaitTasks(), HasLen, 0)
+	c.Assert(tValidateSnap1.WaitTasks(), DeepEquals, []*state.Task{
+		tDownloadSnap1,
+	})
+	c.Assert(tInstallSnap1.WaitTasks(), DeepEquals, []*state.Task{
+		tValidateSnap1,
+		tValidateSnap2,
+		// wait for recovery system to be created
+		tCreateRecovery,
+		// and then finalized
+		tFinalizeRecovery,
+	})
+	c.Assert(tInstallSnap2.WaitTasks(), DeepEquals, []*state.Task{
+		tValidateSnap2,
+		// previous install chain
+		tInstallSnap1,
+	})
+	c.Assert(tCreateRecovery.WaitTasks(), DeepEquals, []*state.Task{
+		// last snap of the download chain
+		tValidateSnap2,
+	})
+	c.Assert(tFinalizeRecovery.WaitTasks(), DeepEquals, []*state.Task{
+		// recovery system being created
+		tCreateRecovery,
+		// last snap of the download chain (added later)
+		tValidateSnap2,
+	})
+
+	c.Assert(tSetModel.Kind(), Equals, "set-model")
+	c.Assert(tSetModel.Summary(), Equals, "Set new model assertion")
```

Comment:
we should check about  recovery-system-setup-task , no?

Reply:
added